Add shuffle Presto function

darrenfu opened the pull request:
`shuffle(x) -> array`

> Generate a random permutation of the given array x.

For example:
```
SELECT shuffle(cast(array[1,2,2,3,4,null,5, null] as array(int))) -- [null, 1, 3, 4, 2, 5, 2, null] or any other permutation
SELECT shuffle(cast(array[null, null] as array(int))) -- [null, null]
SELECT shuffle(array['a', 'a', 'a']) -- ['a', 'a', 'a']
SELECT shuffle(cast(null as array(int))) -- null
```

NOTE: this is a non-deterministic function, hence, it may return different results for the same input even when input is a constant or dictionary encoded vector.

darrenfu commented on velox/vector/tests/utils/VectorTestBase.cpp:
```diff
@@ -127,6 +129,23 @@ void assertEqualVectors(
+  std::multiset<uint64_t> actualSet;
+  for (auto i = 0; i < expected->size(); i++) {
+    expectedSet.insert(expected->hashValueAt(i));
+    actualSet.insert(actual->hashValueAt(i));
```
Added a new test utility function `assertVectorHashEqualsIgnoreOrder()` to compare all hash values ignoring the element order in the vector. This is super useful for skipping-order two vector comparison.
My original plan is to call functions like `valueAt()`, but it is not available in `BaseVector` class.

mbasmanova commented on velox/docs/functions/coverage.rst:
```diff
@@ -216,7 +216,7 @@ Here is a list of all scalar and aggregate Presto functions with functions that
     :func:`array_join`                        :func:`e`                                 json_size                                 sha1                                      st_y                                          :func:`checksum`
     :func:`array_max`                         :func:`element_at`                        :func:`least`                             :func:`sha256`                            st_ymax                                       classification_fall_out
     :func:`array_min`                         :func:`empty_approx_set`                  :func:`length`                            :func:`sha512`                            st_ymin                                       classification_miss_rate
-    array_normalize                           enum_key                                  levenshtein_distance                      shuffle                                   :func:`strpos`                                classification_precision
```
This change should not be part of this PR. After the PR lands, we'll need to update coverage maps separately following instructions in velox/functions/prestosql/coverage/README.md

mbasmanova commented on velox/functions/prestosql/registration/ArrayFunctionsRegistration.cpp:
```diff
@@ -100,5 +108,16 @@ void registerArrayFunctions() {
   registerArrayCombinationsFunctions<Timestamp>();
   registerArrayCombinationsFunctions<Date>();
+
+  registerArrayShuffleFunctions<int8_t>();
```
Can we add support for all input types, e.g. array(array(x)), array(map(k, v)), etc?

darrenfu replied:
I will leave those nested complex types as a TODO item for now because 1) need to change the simple api code; 2) there are more UDFs to support generic complex types. See https://github.com/facebookincubator/velox/issues/3484

mbasmanova commented on velox/functions/prestosql/ArrayFunctions.h:
```diff
@@ -316,4 +318,40 @@ struct ArraySumFunction {
+/// Function Signature: shuffle(x) -> array
+/// Generate a random permutation of the given array x.
+template <typename TExecCtx, typename T>
+struct ArrayShuffleFunction {
```
I feel this function would need to be implemented as a vector function to (1) add support for all types of input including arrays of arrays, maps and structs; (2) avoid copying the input elements.

darrenfu replied:
My feeling is the simple API can support nested complex types, will discuss with @laithsakka on this requirement. There are more UDFs like this one.

For scalar types, shall we stick to simple API? And we can add support for complex types down the line.

mbasmanova commented on velox/functions/prestosql/ArrayFunctions.h:
```diff
@@ -316,4 +318,40 @@ struct ArraySumFunction {
+      const arg_type<velox::Array<T>>& inputArray) {
+    int32_t length = inputArray.size();
+    int32_t indices[length];
+    for (int32_t i = 0; i < length; i++) {
```
use std::iota

mbasmanova commented on velox/functions/prestosql/ArrayFunctions.h:
```diff
@@ -316,4 +318,40 @@ struct ArraySumFunction {
+    for (int32_t i = 0; i < length; i++) {
+      indices[i] = i;
+    }
+    // Fisher-Yates shuffle
```
Any particular reason not to use std::shuffle?

mbasmanova commented on velox/functions/prestosql/ArrayFunctions.h:
```diff
@@ -316,4 +318,40 @@ struct ArraySumFunction {
+      const auto& item = inputArray[indices[i]];
+      if (item.has_value()) {
+        if constexpr (std::is_same_v<T, Varchar>) {
+          out.add_item().setNoCopy(*item);
```
same comment as in the other PR

darrenfu replied:
Like the other PR, l added a TODO to enable zero copy in the future.

mbasmanova commented on velox/functions/prestosql/tests/ArrayShuffleTest.cpp:
```diff
@@ -0,0 +1,255 @@
+
+namespace {
+
+// Class to test the shuffle operator.
```
This comment is redundant. Please, remove.

mbasmanova reviewed (commented):
@darrenfu Thank you for working on adding shuffle function. On top of regular usage, it could help enhance testing for array_sort function by allowing to generate random input then evaluate array_sort(x) and array_sort(shuffle(x)) and assert the results are the same.

CC: @kagamiori

darrenfu commented on velox/functions/prestosql/tests/ArrayShuffleTest.cpp:
```diff
@@ -0,0 +1,310 @@
+  /// verifying the shuffle's (1) randomness and (2) and correctness. However,
+  /// it doesn't guarantee the uniformity. We will avoid over testing with
+  /// leveraging standard deviation, etc.
+  void runTestKTimes(
+      const int32_t t = kNumOfEvalTimes,
+      const int32_t n = kNums) {
```
Add another utility function to test if a shuffle array will be different from others most time.

darrenfu reviewed (commented):
Rewrite unit tests to better test the randomness and correctness of shuffle.

mbasmanova commented:
@darrenfu I'm concerned about not adding support for all types and not implementing this as zero-copy for strings and complex types. What's preventing us from implementing this properly from the start, perhaps, using Vector function?

darrenfu commented:
> @darrenfu Thank you for working on adding shuffle function. On top of regular usage, it could help enhance testing for array_sort function by allowing to generate random input then evaluate array_sort(x) and array_sort(shuffle(x)) and assert the results are the same.

Good callout, @mbasmanova! Apart from the randomness test you mentioned, here I run `shuffle(x)` 100 times with a ranged-based input {0, 1, ..., 9}, and assert:
1. randomness: if the output shuffle arrays will be different at least 70% times in total. (actually, in this case, it should be > 99%)
2. correctness: if it's a permutation of the origin array. I'm using `std::is_permutation()` whose time complexity is O(N^2), a bit slower. Since it's unit test, I think it'll be fine.

This borrows some ideas from Presto java version: https://github.com/prestodb/presto/blob/master/presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java#L4322. Let me know if the new tests suffices.

darrenfu commented on velox/functions/prestosql/tests/ArrayShuffleTest.cpp:
```diff
@@ -0,0 +1,387 @@
+        {row3},
+        {row4}};
+    // TODO: to support T=std::vector<std::optional<int64_t>>
+//    testExpr(expected, "shuffle(C0)", {actual});
```
TODO: to test nested array case.

darrenfu commented on velox/functions/prestosql/ArrayShuffle.cpp:
```diff
@@ -0,0 +1,258 @@
+      localResult =
+          BaseVector::wrapInConstant(rows.size(), flatIndex, localResult);
+    } else {
+      localResult = applyFlat(rows, arg->as<ArrayVector>(), context);
```
TODO: need to support dictionary encoding to deal with current CI fuzzer failure.

darrenfu commented:
The CI failure is caused by a fuzzer bug: https://github.com/facebookincubator/velox/issues/3603

mbasmanova commented on velox/functions/prestosql/ArrayShuffle.cpp:
```diff
@@ -0,0 +1,167 @@
+#include "velox/expression/EvalCtx.h"
+#include "velox/expression/Expr.h"
+#include "velox/expression/VectorFunction.h"
+#include "velox/expression/VectorReaders.h"
```
Is this header needed?

mbasmanova commented on velox/functions/prestosql/ArrayShuffle.cpp:
```diff
@@ -0,0 +1,167 @@
+template <TypeKind T>
+class ArrayShuffleFunction : public exec::VectorFunction {
+ public:
+  ArrayShuffleFunction() {}
```
This is not needed as compiler will generate this automatically.

mbasmanova commented on velox/functions/prestosql/ArrayShuffle.cpp:
```diff
@@ -0,0 +1,167 @@
+namespace {
+// See documentation at
+// https://prestodb.io/docs/current/functions/array.html#shuffle
+template <TypeKind T>
```
Why do we need this template parameter?

mbasmanova commented on velox/functions/prestosql/ArrayShuffle.cpp:
```diff
@@ -0,0 +1,167 @@
+      const auto& flatArray = constantArray->valueVector();
+      const auto flatIndex = constantArray->index();
+
+      exec::LocalSelectivityVector singleRow(context, flatIndex + 1);
```
singleRow is not guaranteed to be initialized to any particular value, hence, it necessary to explicitly deselect everything.

```
singleRow->clearAll();
```

mbasmanova commented on velox/functions/prestosql/ArrayShuffle.cpp:
```diff
@@ -0,0 +1,167 @@
+    } else { // dictionary encoding
+      exec::LocalDecodedVector decodedVectorHolder(context, *arg, rows);
+      localResult = applyFlat(
+          rows, decodedVectorHolder->base()->as<ArrayVector>(), context);
```
This doesn't look right as 'rows' refer to top-level vector, but applyFlat function receives the base vector. For example, top-level vector may have rows 0, 1, ....10, but the the base vector may have only 2 rows: 0, 1. Accesssing rows 2,...10 will cause an error.

darrenfu replied:
Thx for pointing out. I should merge the flat/dict paths, and use `LocalDecodedVector` to make the encoding handling transparent.

mbasmanova commented on velox/functions/prestosql/ArrayShuffle.cpp:
```diff
@@ -0,0 +1,167 @@
+        inputArray->offsets(),
+        inputArray->sizes(),
+        resultElements,
+        inputArray->getNullCount());
```
Do not pass null count. This property is not maintained properly and should not be used.

darrenfu replied:
Created an issue https://github.com/facebookincubator/velox/issues/3641 to track the same issues in other udfs.

mbasmanova commented on velox/functions/prestosql/ArrayShuffle.cpp:
```diff
@@ -0,0 +1,167 @@
+      const SelectivityVector& rows,
+      const ArrayVector* inputArray,
+      exec::EvalCtx& context,
+      VectorPtr& resultElements) const {
```
The signature of this method would be clearer resultElements was returned via 'return'.

```
VectorPtr shuffleElementIndices(rows, inputArray, context)
```

mbasmanova commented on velox/functions/prestosql/ArrayShuffle.cpp:
```diff
@@ -0,0 +1,167 @@
+      const auto size = inputArray->sizeAt(row);
+      const auto offset = inputArray->offsetAt(row);
+
+      std::mt19937 randGen(std::random_device{}());
```
Can we initialize this variable outside of the loop and reuse?

mbasmanova commented on velox/functions/prestosql/ArrayShuffle.cpp:
```diff
@@ -0,0 +1,167 @@
+};
+
+// Validate number of parameters and types.
+void validateType(const std::vector<exec::VectorFunctionArg>& inputArgs) {
```
This validation is redundant. I suggest to remove it.

mbasmanova commented on velox/functions/prestosql/ArrayShuffle.cpp:
```diff
@@ -0,0 +1,167 @@
+
+// Create function template based on type.
+template <TypeKind kind>
+std::shared_ptr<exec::VectorFunction> createTyped(
```
It doesn't look like we need to make a template.

mbasmanova reviewed (commented):
@darrenfu Some comments.

darrenfu commented:
Per discussion with @mbasmanova, I reimplemented the shuffle vector function to cover multi level encodings especially dict encoding.

**Context**: for deterministic single-arg functions, expression evaluation will peel off encodings and guarantee that we will only see flat or constant inputs. But this guarantee does not exist for non-deterministic functions like shuffle(). Hence, we need to use DecodedVector to decode and get the base ArrayVector's elements, and apply the shuffle on each row's indices to construct a new indices array.

Added constant and dict encoding unit test cases as well.

We expect only fuzzer failures like "Different results at idx ...", which will be tracked in https://github.com/facebookincubator/velox/issues/3603

mbasmanova commented on velox/functions/prestosql/ArraySort.cpp:
```diff
@@ -165,8 +165,6 @@ class ArraySortFunction : public exec::VectorFunction {
   /// and 'offsets' vectors that control where output arrays start and end
   /// remain the same in the output ArrayVector.
 
-  ArraySortFunction() {}
```
nice cleanup, but doesn't belong to this PR

mbasmanova commented on velox/functions/prestosql/ArrayShuffle.cpp:
```diff
@@ -0,0 +1,144 @@
+      localResult =
+          BaseVector::wrapInConstant(rows.size(), flatIndex, localResult);
+    } else {
+      // For deterministic single-arg functions, expression evaluation will peel
```
Nice comment.

mbasmanova commented on velox/functions/prestosql/ArrayShuffle.cpp:
```diff
@@ -0,0 +1,144 @@
+  }
+
+ private:
+  VectorPtr shuffleDecodedElementIndices(
```
This name is a bit confusing. It sounds like it takes 'decoded elements' and shuffles them, but this is not what this method does.

darrenfu replied:
Let's just rollback to `applyFlat()` to be consistent with other UDFs.

mbasmanova commented on velox/functions/prestosql/ArrayShuffle.cpp:
```diff
@@ -0,0 +1,144 @@
+      exec::EvalCtx& context) const {
+    exec::LocalDecodedVector localDecodedVector(context, *inputVector, rows);
+    const auto& decodedArray = *localDecodedVector.get();
+    auto flatBaseArray = decodedArray.base()->as<ArrayVector>();
```
nit: consider renaming flatBaseArray and baseArrayElements to arrayVector and elementsVector for clarity.

mbasmanova commented on velox/functions/prestosql/ArrayShuffle.cpp:
```diff
@@ -0,0 +1,144 @@
+    auto flatBaseArray = decodedArray.base()->as<ArrayVector>();
+    auto baseArrayElements = flatBaseArray->elements();
+
+    vector_size_t sizeOfNewIndices = 0;
```
nit: consider renaming to numElements since this is the total number of elements across all arrays in 'rows'.

mbasmanova commented on velox/functions/prestosql/ArrayShuffle.cpp:
```diff
@@ -0,0 +1,144 @@
+      sizeOfNewIndices += size;
+    });
+
+    // Allocate new vectors for shuffledIndices.
```
typos: Allocate new buffer to hold shuffled indices.

mbasmanova commented on velox/functions/prestosql/ArrayShuffle.cpp:
```diff
@@ -0,0 +1,144 @@
+    BufferPtr shuffledIndices =
+        allocateIndices(sizeOfNewIndices, context.pool());
+    BufferPtr offsets = allocateOffsets(rows.size(), context.pool());
+    BufferPtr sizes = allocateOffsets(rows.size(), context.pool());
```
allocateSizes

darrenfu replied:
nice catch!

mbasmanova commented on velox/functions/prestosql/ArrayShuffle.cpp:
```diff
@@ -0,0 +1,144 @@
+    vector_size_t indexCursor = 0;
+    std::mt19937 randGen(std::random_device{}());
+    context.applyToSelectedNoThrow(rows, [&](auto row) {
+      vector_size_t baseIndex = decodedArray.index(row);
```
nit: baseIndex -> baseRow or arrayRow for clarity

```
auto size = arrayVector->sizeAt(arrayRow);
```

mbasmanova commented on velox/functions/prestosql/ArrayShuffle.cpp:
```diff
@@ -0,0 +1,144 @@
+    vector_size_t* rawOffsets = offsets->asMutable<vector_size_t>();
+    vector_size_t* rawSizes = sizes->asMutable<vector_size_t>();
+
+    vector_size_t indexCursor = 0;
```
consider renaming indexCursor -> newOffset

mbasmanova commented on velox/functions/prestosql/ArrayShuffle.cpp:
```diff
@@ -0,0 +1,144 @@
+    });
+
+    auto resultElements = BaseVector::wrapInDictionary(
+        BufferPtr(nullptr),
```
just use nullptr

mbasmanova commented on velox/functions/prestosql/ArrayShuffle.cpp:
```diff
@@ -0,0 +1,144 @@
+  }
+};
+
+// Define function signature.
```
This comment is redundant. Let's remove.

mbasmanova reviewed (commented):
@darrenfu Darren, thank you for iterating on this PR. shuffle turns out to be a non-trivial function to implement. The code looks good to me, but the test seems overly complicated. I didn't have time to read it closely, but given that the function implementation doesn't depend on the type of the array elements, it seems redundant to test different element types.

darrenfu commented:
> @darrenfu Darren, thank you for iterating on this PR. shuffle turns out to be a non-trivial function to implement. The code looks good to me, but the test seems overly complicated. I didn't have time to read it closely, but given that the function implementation doesn't depend on the type of the array elements, it seems redundant to test different element types.

@mbasmanova - like other UDF's tests, I've removed all type related unit test cases. Let me know if you have any concerns.

mbasmanova commented:
@darrenfu There are CI failures. Would you take a look?

darrenfu commented:
> @darrenfu There are CI failures. Would you take a look?

Yeah, it's an expected behavior. Now I just rebased from main branch to include https://github.com/facebookincubator/velox/pull/3633 to skip non-deterministic udf verification on the fuzzer side. The ci failure should go away.

mbasmanova commented on velox/functions/prestosql/tests/ArrayShuffleTest.cpp:
```diff
@@ -0,0 +1,325 @@
+  }
+
+  /// To test a shuffle, we will do the following steps:
+  /// 1. run the test for T times with range-based input {0, 1, ..., n-1},
```
"evaluate shuffle T times on sequential input [0, 1, 2,... n-1]; verify that X% of the results are unique."

mbasmanova commented on velox/functions/prestosql/tests/ArrayShuffleTest.cpp:
```diff
@@ -0,0 +1,325 @@
+  ///    which ensures the shuffle will not miss any existing value or
+  ///    generate any unexpected values.
+  /// NOTE:
+  /// The combination of tests above is a naive but straightforward way of
```
drop: "naive but"

mbasmanova commented on velox/functions/prestosql/tests/ArrayShuffleTest.cpp:
```diff
@@ -0,0 +1,325 @@
+  /// NOTE:
+  /// The combination of tests above is a naive but straightforward way of
+  /// verifying the shuffle's (1) randomness and (2) and correctness. However,
+  /// it doesn't guarantee the uniformity. We will avoid over testing with
```
Drop "We will avoid ..."

mbasmanova commented on velox/functions/prestosql/tests/ArrayShuffleTest.cpp:
```diff
@@ -0,0 +1,325 @@
+  /// it doesn't guarantee the uniformity. We will avoid over testing with
+  /// leveraging standard deviation, etc.
+  void testRandomness(
+      const int32_t t = kNumOfEvalTimes,
```
The callers never specify t or n parameters. Let's remove these.

mbasmanova commented on velox/functions/prestosql/tests/ArrayShuffleTest.cpp:
```diff
@@ -0,0 +1,325 @@
+  void testRandomness(
+      const int32_t t = kNumOfEvalTimes,
+      const int32_t n = kNums) {
+    // generate a range-based array {0, 1, ..., n-1} and shuffle it
```
Make sure all comments start with a capital letter and end with a period.

mbasmanova commented on velox/functions/prestosql/tests/ArrayShuffleTest.cpp:
```diff
@@ -0,0 +1,325 @@
+      const int32_t t = kNumOfEvalTimes,
+      const int32_t n = kNums) {
+    // generate a range-based array {0, 1, ..., n-1} and shuffle it
+    std::vector<int32_t> actual(n);
```
actual -> input

expected variable can be removed. We can just assert that shuffle results are a permutation of 'input'.

mbasmanova commented on velox/functions/prestosql/tests/ArrayShuffleTest.cpp:
```diff
@@ -0,0 +1,325 @@
+    folly::F14FastSet<std::vector<int32_t>> distinctResults;
+
+    for (auto i = 0; i < t; i++) {
+      auto arrayVectorResult =
```
arrayVectorResult -> result

mbasmanova commented on velox/functions/prestosql/tests/ArrayShuffleTest.cpp:
```diff
@@ -0,0 +1,325 @@
+          evaluate<ArrayVector>("shuffle(C0)", makeRowVector({actualVector}));
+      // decode element values from output ArrayVector
+      DecodedVector decoded;
+      exec::VectorReader<Any> reader(decode(decoded, *arrayVectorResult));
```
Any -> Array<int32_t> ?

mbasmanova commented on velox/functions/prestosql/tests/ArrayShuffleTest.cpp:
```diff
@@ -0,0 +1,325 @@
+      DecodedVector decoded;
+      exec::VectorReader<Any> reader(decode(decoded, *arrayVectorResult));
+      auto arrayReader = reader[0].castTo<Array<int32_t>>();
+      std::vector<int32_t> shuffledResult(kNums);
```
Any particular reason not to use VectorReader::materialize?

darrenfu replied:
Thanks for raising it. 
One gotcha: since my input and output arrays are both null-free, I found instead of calling `reader[0]`, I need to call `reader.readNullFree(0)` to force it return a `std::vector<int32_t>`.

mbasmanova commented on velox/functions/prestosql/tests/ArrayShuffleTest.cpp:
```diff
@@ -0,0 +1,325 @@
+        << numOfDistincts;
+  }
+
+  void testBigintNullable() {
```
This method is used only once. Consider folding its logic into the caller. Ditto other similar methods.

mbasmanova commented on velox/functions/prestosql/tests/ArrayShuffleTest.cpp:
```diff
@@ -0,0 +1,325 @@
+             std::nullopt,
+             std::nullopt,
+         }});
+    std::vector<std::vector<std::optional<int64_t>>> expected{
```
Why can't we use `auto expected = actual`"?

Also, can we modify testIsPermutation to `testShuffle(input)` and have it take a single ArrayVector as input, evaluate shuffle on it and verify that result is a permutation of input?

mbasmanova commented on velox/functions/prestosql/tests/ArrayShuffleTest.cpp:
```diff
@@ -0,0 +1,325 @@
+         std::numeric_limits<int64_t>::max(),
+         std::numeric_limits<int64_t>::max(),
+         std::numeric_limits<int64_t>::max()}};
+    testIsPermutation(expected, "shuffle(C0)", {actual});
```
This can be simplified to `testShuffle(input)` (after renaming actual to input).

mbasmanova reviewed (commented):
@darrenfu The implementation of the shuffle function looks great. I have some comments about the test.

darrenfu commented on velox/functions/prestosql/ArrayShuffle.cpp:
```diff
@@ -0,0 +1,112 @@
+    // inputs. But this guarantee does not exist for non-deterministic
+    // functions like shuffle(). Hence, we need to use DecodedVector to handle
+    // all the possible dict encodings including constant, flat or dictionary.
+    exec::LocalDecodedVector localDecodedVector(context, *args[0], rows);
```
Sg, from src, I can tell that `DecodedArgs` maintains a list of `LocalDecodedVector` based on # of input args, which is neater than instantiating `LocalDecodedVector` for a specific argument.

darrenfu commented:
@mbasmanova thanks for your scrutiny on the unit test class. I addressed all your comments, and added  three more test cases to test the shuffle randomness besides the shuffle correctness (is-a-permutation). Please see the details in: `testShuffleRandomness()`

mbasmanova commented on velox/functions/prestosql/tests/ArrayShuffleTest.cpp:
```diff
@@ -0,0 +1,316 @@
+  constexpr static const int32_t kNums = 10;
+  constexpr static const double kUniquenessRate = .7;
+
+  template <typename T, bool returnsOptionalValues = false>
```
All the calls to testShuffle specify returnsOptionalValues as true. Hence, this template parameter is not needed. Let's remove it.

darrenfu replied:
Makes sense. 
FYI: my original idea is to have only one testShuffle() function instead of two functions: testShuffle() + testShuffleRandomness(). Somehow, I need to deal with build errors caused by 1) having std::optional and nested std::vector in a folly::F14FastSet and 2) calculating the right sizes & thresholds for different shape of input array cases. Then, I ended up choosing to separate into two functions to keep brevity.

mbasmanova commented on velox/functions/prestosql/tests/ArrayShuffleTest.cpp:
```diff
@@ -0,0 +1,316 @@
+  template <typename T, bool returnsOptionalValues = false>
+  void testShuffle(const VectorPtr& input) {
+    DecodedVector decodedExpected;
+    decodedExpected.decode(*input.get());
```
This line can be combined with the previous:

```
DecodedVector decodedExpected(*input.get());
```

mbasmanova commented on velox/functions/prestosql/tests/ArrayShuffleTest.cpp:
```diff
@@ -0,0 +1,316 @@
+    // Validate each row from the actual decoded ArrayVector is a permutation
+    // of the corresponding row from the expected decoded ArrayVector.
+    DecodedVector decodedActual;
+    decodedActual.decode(*actualVector.get());
```
ditto

mbasmanova commented on velox/functions/prestosql/tests/ArrayShuffleTest.cpp:
```diff
@@ -0,0 +1,316 @@
+
+    for (auto i = 0; i < input->size(); i++) {
+      // Must materialize into std::vector, otherwise it will throw error
+      // because ArrayView doesn't support std::is_permutation() yet.
```
@laithsakka Would it make sense to add support for std::is_permutation() and similar to Array reader?

darrenfu replied:
I agree. I updated the origin enhancement task with this requirement:
https://github.com/facebookincubator/velox/issues/3426

mbasmanova commented on velox/functions/prestosql/tests/ArrayShuffleTest.cpp:
```diff
@@ -0,0 +1,316 @@
+      // Must materialize into std::vector, otherwise it will throw error
+      // because ArrayView doesn't support std::is_permutation() yet.
+      auto actualArray =
+          read<T, returnsOptionalValues>(readerActual, i).materialize();
```
Since returnsOptionalValues is always true, this can be simplified to readerActual[i] and read function may be deleted.

mbasmanova commented on velox/functions/prestosql/tests/ArrayShuffleTest.cpp:
```diff
@@ -0,0 +1,316 @@
+      auto expectedArray =
+          read<T, returnsOptionalValues>(readerExpected, i).materialize();
+
+      // Assert the two vectors containing the same elements (ignoring order).
```
typo: containing -> contain

mbasmanova commented on velox/functions/prestosql/tests/ArrayShuffleTest.cpp:
```diff
@@ -0,0 +1,316 @@
+      ASSERT_TRUE(std::is_permutation(
+          actualArray.begin(), actualArray.end(), expectedArray.begin()))
+          << "Actual array " << printArray(actualArray) << " at " << i
+          << " must produce a permutation of expected array "
```
typo: produce -> be

mbasmanova commented on velox/functions/prestosql/tests/ArrayShuffleTest.cpp:
```diff
@@ -0,0 +1,316 @@
+  /// ArrayVector (with and identical elements: N) and a t-size indices.
+  VectorPtr generateVector(
+      const VectorEncoding::Simple& encoding,
+      const int32_t n = kNums,
```
Like mentioned earlier n and t parameters are never used. Let's remove.

mbasmanova commented on velox/functions/prestosql/tests/ArrayShuffleTest.cpp:
```diff
@@ -0,0 +1,316 @@
+namespace {
+class ArrayShuffleTest : public FunctionBaseTest {
+ protected:
+  constexpr static const int32_t kNumShuffleTimes = 100;
```
These constants are used only in one method. Let's move them there.

mbasmanova commented on velox/functions/prestosql/tests/ArrayShuffleTest.cpp:
```diff
@@ -0,0 +1,316 @@
+    std::vector<int32_t> inputData(n);
+    std::iota(inputData.begin(), inputData.end(), 0);
+
+    if (encoding == VectorEncoding::Simple::FLAT) {
```
nit: consider using switch

mbasmanova commented on velox/functions/prestosql/tests/ArrayShuffleTest.cpp:
```diff
@@ -0,0 +1,316 @@
+    std::iota(inputData.begin(), inputData.end(), 0);
+
+    if (encoding == VectorEncoding::Simple::FLAT) {
+      std::vector<std::vector<int32_t>> flatData(t);
```
Using single-letter variables is an anti pattern because it makes code hard to read. Can we use descriptive names for t and n?

mbasmanova commented on velox/functions/prestosql/tests/ArrayShuffleTest.cpp:
```diff
@@ -0,0 +1,316 @@
+      return makeArrayVector<int32_t>(flatData);
+    }
+    if (encoding == VectorEncoding::Simple::CONSTANT) {
+      vector_size_t size = t;
```
no need to introduce temp variable 'size'; let's use t directly

mbasmanova commented on velox/functions/prestosql/tests/ArrayShuffleTest.cpp:
```diff
@@ -0,0 +1,316 @@
+      auto indices = makeIndices(indicesData);
+      return wrapInDictionary(indices, base);
+    }
+    // Never reach here.
```
use VELOX_UNREACHABLE("Unsupported encoding: {}", mapSimpleToName(encoding))

mbasmanova commented on velox/functions/prestosql/tests/ArrayShuffleTest.cpp:
```diff
@@ -0,0 +1,316 @@
+};
+} // namespace
+
+TEST_F(ArrayShuffleTest, normalArrays) {
```
It is not clear what "normal" means here. Maybe, bigintArrays?

mbasmanova commented on velox/functions/prestosql/tests/ArrayShuffleTest.cpp:
```diff
@@ -0,0 +1,316 @@
+
+TEST_F(ArrayShuffleTest, constantEncoding) {
+  vector_size_t size = 100;
+  // Considering the cases of empty array, nullable array, duplicate-element
```
typo: "Considering the cases" -> "Test empty array, array with null element and array with duplicate elements."

mbasmanova commented on velox/functions/prestosql/tests/ArrayShuffleTest.cpp:
```diff
@@ -0,0 +1,316 @@
+}
+
+TEST_F(ArrayShuffleTest, dictEncoding) {
+  // Considering the case of repeated elements: {1,2,3} x 3, {4,5} x 2.
```
This use case doesn't appear to be tested. Would you double check?

darrenfu replied:
What do you mean "doesn't appear to be tested"? 
Right now, I'm using std::is_permutation() to test if the shuffled output is a permutation of the origin input in test_shuffle() for the **correctness purposes**.
Regarding the **randomness test**, you will need to look at the test case: `dictEncodingRandomness`. It will generate a dictionary with size=200 base data and size=100 indices. The dictionary values will all be identical: {0,1,2,...,9}. And it will test if the shuffled output contains at least 7 distinct permutations (e.g. {1,2,3},{1,2,3},... => {1,3,2}, {3,2,1}). Putting in another word, if it shuffles only once for those repeated dict elements, it will yield one permutation (e.g. {1,2,3} => {1,3,2}). Does this new test case cover your question?

mbasmanova commented on velox/functions/prestosql/tests/ArrayShuffleTest.cpp:
```diff
@@ -0,0 +1,316 @@
+       {1, 2, 3},
+       {4, 5, std::nullopt}});
+  // Considering the case of indices filtering (filter out element at index 0)
+  std::vector<vector_size_t> indicesArray({3, 4, 1, 2});
```
inline indicesArray variable for readability:

```
auto indices = makeIndices({3, 4, 1, 2});
```

mbasmanova commented on velox/functions/prestosql/tests/ArrayShuffleTest.cpp:
```diff
@@ -0,0 +1,316 @@
+
+TEST_F(ArrayShuffleTest, flatEncodingRandomness) {
+  auto inputVector = generateVector(VectorEncoding::Simple::FLAT);
+  testShuffleRandomness(inputVector);
```
testShuffleRandomness is always used together with generateVector. Consider folding generateVector logic into testShuffleRandomness and simplify the callers:

```
testShuffleRandomness(VectorEncoding::Simple::FLAT);
```

mbasmanova reviewed (commented):
@darrenfu Some additional comments on the test. I'm seeing a number of method with more parameters then needed / used.

darrenfu commented:
With the latest test cases: `constantEncoding` and `constantEncodingRandomness`, if revert to special handling on constant encoding, (cc0cbd7dd91188db6b6bce665f96c80e997c0a54), it will report the following error. It means these two test cases are necessary, and we shouldn't do any special handling on constant encoding as in other array UDFs. 
```
unknown file: Failure
C++ exception with description "Exception: VeloxRuntimeError
Error Source: RUNTIME
Error Code: INVALID_STATE
Reason: Expression evaluation result is not of expected type: shuffle(C0) -> CONSTANT vector of type ARRAY<BIGINT>
Retriable: False
Expression: castedResult
Function: castEvaluateResult
File: /Users/dofu/git/velox/./velox/functions/prestosql/tests/utils/FunctionBaseTest.h
Line: 274
```

Thanks @mbasmanova to call it out.

mbasmanova reviewed (commented):
@darrenfu Thank you, Darren. Consider updating PR description to clarify that shuffle is a non-deterministic function, hence, it may return different results for the same input even when input is a constant or dictionary encoded vector.
